apoc.refactor.mergeNodes throws IndexEntryConflictException when merged nodes have relationships with UNIQUE constraint

I discovered this when trying to merge nodes using `apoc.refactor.mergeNodes` procedure.
If a node I'm merging has a relationship of type with a unique constraint, then the merger operation fails with `IndexEntryConflictException`.

**Neo4j Version:** 5.25.1 + APOC 5.25.1
**Operating System:** macOS 15.1.1
**API:** Docker

### Steps to reproduce
1. Start with an empty DB.
2. Run the following to create the constraint and populate the DB with some nodes and relationships:
```cypher
CREATE CONSTRAINT foo_uniq FOR ()-[r:MY_REL]-() REQUIRE r.foo IS UNIQUE;
CREATE
    (n1 {name: "n1"})-[r1:MY_REL {foo: "a"}]->(n2 {name: "n2"}),
    (n3 {name: "n3"})-[r2:MY_REL {foo: "b"}]->(n4 {name: "n4"});
```
3. Attempt to merge nodes `n1` and `n3`:
```cypher
MATCH (n1 {name: "n1"}), (n3 {name: "n3"})
CALL apoc.refactor.mergeNodes([n1, n3]) YIELD node
FINISH
```

### Expected behavior
Nodes are merged and we're left with:
`(n1)-[r1]->(n2)`
`(n1)-[r2]->(n4)` 

This is what happens if `foo_uniq` constraint is not present.

### Actual behavior
The query returns:
> Failed to invoke procedure `apoc.refactor.mergeNodes`: Caused by: IndexEntryConflictException{propertyValues=( String("b") ), addedEntityId=-1, existingEntityId=1}

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -913,21 +913,22 @@ public class GraphRefactoring {
 
     private void copyRelationships(Node source, Node target, boolean delete, boolean createNewSelfRel) {
         for (Relationship rel : source.getRelationships()) {
-            final var type = rel.getType();
             var startNode = rel.getStartNode();
-            if (startNode.getElementId().equals(source.getElementId())) startNode = target;
             var endNode = rel.getEndNode();
+
+            if (!createNewSelfRel && startNode.getElementId().equals(endNode.getElementId())) continue;
+
+            if (startNode.getElementId().equals(source.getElementId())) startNode = target;
             if (endNode.getElementId().equals(source.getElementId())) endNode = target;
 
+            final var type = rel.getType();
             final var properties = rel.getAllProperties();
 
             // Delete first to avoid breaking constraints.
             if (delete) rel.delete();
 
-            if (createNewSelfRel || !startNode.getElementId().equals(endNode.getElementId())) {
-                final var newRel = startNode.createRelationshipTo(endNode, type);
-                properties.forEach(newRel::setProperty);
-            }
+            final var newRel = startNode.createRelationshipTo(endNode, type);
+            properties.forEach(newRel::setProperty);
         }
     }
 }

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -916,19 +916,21 @@ public class GraphRefactoring {
             var startNode = rel.getStartNode();
             var endNode = rel.getEndNode();
 
-            if (!createNewSelfRel && startNode.getElementId().equals(endNode.getElementId())) continue;
+            if (!createNewSelfRel && startNode.getElementId().equals(endNode.getElementId()))  {
+                if (delete) rel.delete();
+            } else {
+                if (startNode.getElementId().equals(source.getElementId())) startNode = target;
+                if (endNode.getElementId().equals(source.getElementId())) endNode = target;
 
-            if (startNode.getElementId().equals(source.getElementId())) startNode = target;
-            if (endNode.getElementId().equals(source.getElementId())) endNode = target;
+                final var type = rel.getType();
+                final var properties = rel.getAllProperties();
 
-            final var type = rel.getType();
-            final var properties = rel.getAllProperties();
+                // Delete first to avoid breaking constraints.
+                if (delete) rel.delete();
 
-            // Delete first to avoid breaking constraints.
-            if (delete) rel.delete();
-
-            final var newRel = startNode.createRelationshipTo(endNode, type);
-            properties.forEach(newRel::setProperty);
+                final var newRel = startNode.createRelationshipTo(endNode, type);
+                properties.forEach(newRel::setProperty);
+            }
         }
     }
 }

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -916,7 +916,7 @@ public class GraphRefactoring {
             var startNode = rel.getStartNode();
             var endNode = rel.getEndNode();
 
-            if (!createNewSelfRel && startNode.getElementId().equals(endNode.getElementId()))  {
+            if (!createNewSelfRel && startNode.getElementId().equals(endNode.getElementId())) {
                 if (delete) rel.delete();
             } else {
                 if (startNode.getElementId().equals(source.getElementId())) startNode = target;

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -477,16 +477,44 @@ public class GraphRefactoring {
     @Description("Redirects the given `RELATIONSHIP` to the given end `NODE`.")
     public Stream<UpdatedRelationshipResult> to(
             @Name(value = "rel", description = "The relationship to redirect.") Relationship rel,
-            @Name(value = "endNode", description = "The new end node the relationship should point to.") Node newNode) {
+            @Name(value = "endNode", description = "The new end node the relationship should point to.") Node newNode,
+            @Name(
+                            value = "config",
+                            defaultValue = "{}",
+                            description =
+                                    """
+                {
+                    failOnErrors = false :: BOOLEAN
+                }
+                """)
+                    Map<String, Object> config) {
+        if (config == null) config = Map.of();
         if (rel == null || newNode == null) return Stream.empty();
-        UpdatedRelationshipResult result = new UpdatedRelationshipResult(rel.getId());
+
+        final var failOnErrors = Boolean.TRUE.equals(config.getOrDefault("failOnErrors", false));
+        final var result = new UpdatedRelationshipResult(rel.getId());
+
         try {
-            Relationship newRel = rel.getStartNode().createRelationshipTo(newNode, rel.getType());
-            copyProperties(rel, newRel);
+            final var type = rel.getType();
+            final var properties = rel.getAllProperties();
+            final var startNode = rel.getStartNode();
+
+            // Delete first to not break constraints.
             rel.delete();
+
+            final var newRel = startNode.createRelationshipTo(newNode, type);
+            properties.forEach(newRel::setProperty);
+
             return Stream.of(result.withOther(newRel));
         } catch (Exception e) {
-            return Stream.of(result.withError(e));
+            if (failOnErrors) {
+                throw e;
+            } else {
+                // Note! We might now have half applied the changes, not sure why we would want to do this instead of
+                // just failing.
+                // I guess it's up to the user to explicitly rollback at this point ¯\_(ツ)_/¯.
+                return Stream.of(result.withError(e));
+            }
         }
     }
 

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -510,8 +510,8 @@ public class GraphRefactoring {
             if (failOnErrors) {
                 throw e;
             } else {
-                // Note! We might now have half applied the changes, not sure why we would want to do this instead of
-                // just failing.
+                // Note! We might now have half applied the changes,
+                // not sure why we would ever want to do this instead of just failing.
                 // I guess it's up to the user to explicitly rollback at this point ¯\_(ツ)_/¯.
                 return Stream.of(result.withError(e));
             }
@@ -555,8 +555,8 @@ public class GraphRefactoring {
             if (failOnErrors) {
                 throw e;
             } else {
-                // Note! We might now have half applied the changes, not sure why we would want to do this instead of
-                // just failing.
+                // Note! We might now have half applied the changes,
+                // not sure why we would ever want to do this instead of just failing.
                 // I guess it's up to the user to explicitly rollback at this point ¯\_(ツ)_/¯.
                 return Stream.of(result.withError(e));
             }
@@ -603,8 +603,8 @@ public class GraphRefactoring {
             if (failOnErrors) {
                 throw e;
             } else {
-                // Note! We might now have half applied the changes, not sure why we would want to do this instead of
-                // just failing.
+                // Note! We might now have half applied the changes,
+                // not sure why we would ever want to do this instead of just failing.
                 // I guess it's up to the user to explicitly rollback at this point ¯\_(ツ)_/¯.
                 return Stream.of(result.withError(e));
             }

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -913,37 +913,21 @@ public class GraphRefactoring {
 
     private void copyRelationships(Node source, Node target, boolean delete, boolean createNewSelfRel) {
         for (Relationship rel : source.getRelationships()) {
-            copyRelationship(rel, source, target, createNewSelfRel);
-            if (delete) rel.delete();
-        }
-    }
+            final var type = rel.getType();
+            var startNode = rel.getStartNode();
+            if (startNode.getElementId().equals(source.getElementId())) startNode = target;
+            var endNode = rel.getEndNode();
+            if (endNode.getElementId().equals(source.getElementId())) endNode = target;
 
-    private Node copyLabels(Node source, Node target) {
-        for (Label label : source.getLabels()) {
-            if (!target.hasLabel(label)) {
-                target.addLabel(label);
+            final var properties = rel.getAllProperties();
+
+            // Delete first to avoid breaking constraints.
+            if (delete) rel.delete();
+
+            if (createNewSelfRel || !startNode.getElementId().equals(endNode.getElementId())) {
+                final var newRel = startNode.createRelationshipTo(endNode, type);
+                properties.forEach(newRel::setProperty);
             }
         }
-        return target;
-    }
-
-    private void copyRelationship(Relationship rel, Node source, Node target, boolean createNewSelfRelf) {
-        Node startNode = rel.getStartNode();
-        Node endNode = rel.getEndNode();
-
-        if (startNode.getElementId().equals(endNode.getElementId()) && !createNewSelfRelf) {
-            return;
-        }
-
-        if (startNode.getElementId().equals(source.getElementId())) {
-            startNode = target;
-        }
-
-        if (endNode.getElementId().equals(source.getElementId())) {
-            endNode = target;
-        }
-
-        Relationship newrel = startNode.createRelationshipTo(endNode, rel.getType());
-        copyProperties(rel, newrel);
     }
 }

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -521,16 +521,45 @@ public class GraphRefactoring {
     @Procedure(name = "apoc.refactor.invert", mode = Mode.WRITE, eager = true)
     @Description("Inverts the direction of the given `RELATIONSHIP`.")
     public Stream<RefactorRelationshipResult> invert(
-            @Name(value = "rel", description = "The relationship to reverse.") Relationship rel) {
+            @Name(value = "rel", description = "The relationship to reverse.") Relationship rel,
+            @Name(
+                            value = "config",
+                            defaultValue = "{}",
+                            description =
+                                    """
+                    {
+                        failOnErrors = false :: BOOLEAN
+                    }
+                    """)
+                    Map<String, Object> config) {
+        if (config == null) config = Map.of();
         if (rel == null) return Stream.empty();
-        RefactorRelationshipResult result = new RefactorRelationshipResult(rel.getId());
+
+        final var failOnErrors = Boolean.TRUE.equals(config.getOrDefault("failOnErrors", false));
+        final var result = new RefactorRelationshipResult(rel.getId());
+
         try {
-            Relationship newRel = rel.getEndNode().createRelationshipTo(rel.getStartNode(), rel.getType());
-            copyProperties(rel, newRel);
+            final var type = rel.getType();
+            final var properties = rel.getAllProperties();
+            final var startNode = rel.getStartNode();
+            final var endNode = rel.getEndNode();
+
+            // Delete first to not break constraints.
             rel.delete();
+
+            final var newRel = endNode.createRelationshipTo(startNode, type);
+            properties.forEach(newRel::setProperty);
+
             return Stream.of(result.withOther(newRel));
         } catch (Exception e) {
-            return Stream.of(result.withError(e));
+            if (failOnErrors) {
+                throw e;
+            } else {
+                // Note! We might now have half applied the changes, not sure why we would want to do this instead of
+                // just failing.
+                // I guess it's up to the user to explicitly rollback at this point ¯\_(ツ)_/¯.
+                return Stream.of(result.withError(e));
+            }
         }
     }
 

--- a/core/src/main/java/apoc/refactor/GraphRefactoring.java
+++ b/core/src/main/java/apoc/refactor/GraphRefactoring.java
@@ -570,16 +570,44 @@ public class GraphRefactoring {
     @Description("Redirects the given `RELATIONSHIP` to the given start `NODE`.")
     public Stream<RefactorRelationshipResult> from(
             @Name(value = "rel", description = "The relationship to redirect.") Relationship rel,
-            @Name(value = "newNode", description = "The node to redirect the given relationship to.") Node newNode) {
+            @Name(value = "newNode", description = "The node to redirect the given relationship to.") Node newNode,
+            @Name(
+                            value = "config",
+                            defaultValue = "{}",
+                            description =
+                                    """
+                    {
+                        failOnErrors = false :: BOOLEAN
+                    }
+                    """)
+                    Map<String, Object> config) {
+        if (config == null) config = Map.of();
         if (rel == null || newNode == null) return Stream.empty();
-        RefactorRelationshipResult result = new RefactorRelationshipResult(rel.getId());
+
+        final var result = new RefactorRelationshipResult(rel.getId());
+        final var failOnErrors = Boolean.TRUE.equals(config.getOrDefault("failOnErrors", false));
+
         try {
-            Relationship newRel = newNode.createRelationshipTo(rel.getEndNode(), rel.getType());
-            copyProperties(rel, newRel);
+            final var type = rel.getType();
+            final var properties = rel.getAllProperties();
+            final var endNode = rel.getEndNode();
+
+            // Delete before setting properties to not break constraints.
             rel.delete();
+
+            final var newRel = newNode.createRelationshipTo(endNode, type);
+            properties.forEach(newRel::setProperty);
+
             return Stream.of(result.withOther(newRel));
         } catch (Exception e) {
-            return Stream.of(result.withError(e));
+            if (failOnErrors) {
+                throw e;
+            } else {
+                // Note! We might now have half applied the changes, not sure why we would want to do this instead of
+                // just failing.
+                // I guess it's up to the user to explicitly rollback at this point ¯\_(ツ)_/¯.
+                return Stream.of(result.withError(e));
+            }
         }
     }
 

--- a/core/src/test/java/apoc/ArgumentDescriptionsTest.java
+++ b/core/src/test/java/apoc/ArgumentDescriptionsTest.java
@@ -20,6 +20,7 @@ package apoc;
 
 import static java.util.Collections.singletonList;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
 
 import apoc.agg.CollAggregation;
 import apoc.agg.Graph;
@@ -329,7 +330,10 @@ public class ArgumentDescriptionsTest {
         System.out.println("common:");
         System.out.println(writer.writeValueAsString(commonJson));
         System.out.println();
+        System.out.println("=".repeat(80));
+        System.out.println();
         System.out.println("cypher %s specific:".formatted(version.versionName));
         System.out.println(writer.writeValueAsString(specificJson));
+        fail("Remove call to generate");
     }
 }

--- a/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
+++ b/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
@@ -1808,6 +1808,88 @@ public class GraphRefactoringTest {
     }
 
     @Test
+    public void refactorFromWithConstraints() {
+        db.executeTransactionally("CREATE CONSTRAINT id_unique FOR ()-[r:R]-() REQUIRE r.id IS UNIQUE");
+        db.executeTransactionally("CREATE (a:A {id:'A'})-[r:R {id:'R'}]->(b:B {id:'B'}), (c:C {id:'C'})");
+        db.executeTransactionally("CALL db.awaitIndexes()");
+
+        try (final var tx = db.beginTx()) {
+            assertThat(tx
+                            .execute(
+                                    "MATCH (a:A)-[r:R]->(), (c:C) CALL apoc.refactor.from(r, c) YIELD error RETURN error")
+                            .stream())
+                    .singleElement(InstanceOfAssertFactories.map(String.class, Object.class))
+                    .containsEntry("error", null);
+            tx.commit();
+        }
+        try (final var tx = db.beginTx()) {
+            final var query =
+                    """
+                    MATCH (a)
+                    OPTIONAL MATCH (a)-[r]->(b)
+                    RETURN
+                      a.id,
+                      CASE WHEN r.id IS NULL THEN 'null' ELSE r.id END AS `r.id`,
+                      CASE WHEN b.id IS NULL THEN 'null' ELSE b.id END AS `b.id`
+                    ORDER BY `a.id`, `r.id`, `b.id`
+                    """;
+            assertThat(tx.execute(query).stream())
+                    .containsExactly(
+                            Map.of("a.id", "A", "r.id", "null", "b.id", "null"),
+                            Map.of("a.id", "B", "r.id", "null", "b.id", "null"),
+                            Map.of("a.id", "C", "r.id", "R", "b.id", "B"));
+            tx.commit();
+        }
+    }
+
+    @Test
+    public void refactorFromWithErrorHandling() {
+        db.executeTransactionally("CREATE (a:A {id:'A'})-[r:R {id:'R'}]->(b:B {id:'B'}), (c:C {id:'C'})");
+
+        final var refactorQuery =
+                """
+                   MATCH (a:A)-[r:R]->(), (c:C)
+                   DELETE r WITH r, c
+                   CALL apoc.refactor.from(r, c, $conf) YIELD error
+                   RETURN error""";
+        try (final var tx = db.beginTx()) {
+            final var params = Map.<String, Object>of("conf", Map.of("failOnErrors", true));
+            assertThatThrownBy(() -> tx.execute(refactorQuery, params).resultAsString())
+                    .hasMessageContaining(
+                            "Failed to invoke procedure `apoc.refactor.from`: Caused by: org.neo4j.internal.kernel.api.exceptions.EntityNotFoundException")
+                    .hasRootCauseExactlyInstanceOf(
+                            org.neo4j.internal.kernel.api.exceptions.EntityNotFoundException.class);
+            tx.rollback();
+        }
+
+        try (final var tx = db.beginTx()) {
+            final var query =
+                    """
+                    MATCH (a)
+                    OPTIONAL MATCH (a)-[r]->(b)
+                    RETURN
+                      a.id,
+                      CASE WHEN r.id IS NULL THEN 'null' ELSE r.id END AS `r.id`,
+                      CASE WHEN b.id IS NULL THEN 'null' ELSE b.id END AS `b.id`
+                    ORDER BY `a.id`, `r.id`, `b.id`
+                    """;
+            assertThat(tx.execute(query).stream())
+                    .containsExactly(
+                            Map.of("a.id", "A", "r.id", "R", "b.id", "B"),
+                            Map.of("a.id", "B", "r.id", "null", "b.id", "null"),
+                            Map.of("a.id", "C", "r.id", "null", "b.id", "null"));
+            tx.commit();
+        }
+
+        try (final var tx = db.beginTx()) {
+            assertThat(tx.execute(refactorQuery, Map.of("conf", Map.of())).stream())
+                    .singleElement(InstanceOfAssertFactories.map(String.class, Object.class))
+                    .hasEntrySatisfying("error", e -> assertThat(e).asString().contains("EntityNotFoundException"));
+            tx.rollback();
+        }
+    }
+
+    @Test
     public void invertWithConstraints() {
         db.executeTransactionally("CREATE CONSTRAINT id_unique FOR ()-[r:R]-() REQUIRE r.id IS UNIQUE");
         db.executeTransactionally("CREATE (a:A {id:'A'})-[r:R {id:'R'}]->(b:B {id:'B'})");

--- a/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
+++ b/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
@@ -28,6 +28,7 @@ import static java.util.Collections.emptyList;
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.singletonList;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.assertj.core.api.InstanceOfAssertFactories.list;
 import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
@@ -54,6 +55,7 @@ import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 import junit.framework.TestCase;
 import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.assertj.core.api.InstanceOfAssertFactories;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Rule;
@@ -1724,6 +1726,85 @@ public class GraphRefactoringTest {
     @Test
     public void issue2797WithCloneSubgraph() {
         issue2797Common(CLONE_SUBGRAPH_QUERY);
+    }
+
+    @Test
+    public void refactorToWithConstraints() {
+        db.executeTransactionally("CREATE CONSTRAINT id_unique FOR ()-[r:R]-() REQUIRE r.id IS UNIQUE");
+        db.executeTransactionally("CREATE (a:A {id:'A'})-[r:R {id:'R'}]->(b:B {id:'B'}), (c:C {id:'C'})");
+        db.executeTransactionally("CALL db.awaitIndexes()");
+
+        try (final var tx = db.beginTx()) {
+            final var refactorToQuery =
+                    "MATCH (a:A)-[r:R]->(), (c:C) CALL apoc.refactor.to(r, c) YIELD error RETURN error";
+            assertThat(tx.execute(refactorToQuery).stream())
+                    .singleElement(InstanceOfAssertFactories.map(String.class, Object.class))
+                    .containsEntry("error", null);
+            tx.commit();
+        }
+        try (final var tx = db.beginTx()) {
+            final var query =
+                    """
+                    MATCH (a)
+                    OPTIONAL MATCH (a)-[r]->(b)
+                    RETURN
+                      a.id,
+                      CASE WHEN r.id IS NULL THEN 'null' ELSE r.id END AS `r.id`,
+                      CASE WHEN b.id IS NULL THEN 'null' ELSE b.id END AS `b.id`
+                    ORDER BY `a.id`, `r.id`, `b.id`
+                    """;
+            assertThat(tx.execute(query).stream())
+                    .containsExactly(
+                            Map.of("a.id", "A", "r.id", "R", "b.id", "C"),
+                            Map.of("a.id", "B", "r.id", "null", "b.id", "null"),
+                            Map.of("a.id", "C", "r.id", "null", "b.id", "null"));
+            tx.commit();
+        }
+    }
+
+    @Test
+    public void refactorToWithErrorHandling() {
+        db.executeTransactionally("CREATE (a:A {id:'A'})-[r:R {id:'R'}]->(b:B {id:'B'}), (c:C {id:'C'})");
+
+        final var refactorQuery =
+                """
+                   MATCH (a:A)-[r:R]->(), (c:C)
+                   DELETE r WITH r, c
+                   CALL apoc.refactor.to(r, c, $conf) YIELD error
+                   RETURN error""";
+        try (final var tx = db.beginTx()) {
+            final var params = Map.<String, Object>of("conf", Map.of("failOnErrors", true));
+            assertThatThrownBy(() -> tx.execute(refactorQuery, params).resultAsString())
+                    .hasMessageContaining(
+                            "Failed to invoke procedure `apoc.refactor.to`: Caused by: org.neo4j.internal.kernel.api.exceptions.EntityNotFoundException");
+            tx.rollback();
+        }
+
+        try (final var tx = db.beginTx()) {
+            final var query =
+                    """
+                    MATCH (a)
+                    OPTIONAL MATCH (a)-[r]->(b)
+                    RETURN
+                      a.id,
+                      CASE WHEN r.id IS NULL THEN 'null' ELSE r.id END AS `r.id`,
+                      CASE WHEN b.id IS NULL THEN 'null' ELSE b.id END AS `b.id`
+                    ORDER BY `a.id`, `r.id`, `b.id`
+                    """;
+            assertThat(tx.execute(query).stream())
+                    .containsExactly(
+                            Map.of("a.id", "A", "r.id", "R", "b.id", "B"),
+                            Map.of("a.id", "B", "r.id", "null", "b.id", "null"),
+                            Map.of("a.id", "C", "r.id", "null", "b.id", "null"));
+            tx.commit();
+        }
+
+        try (final var tx = db.beginTx()) {
+            assertThat(tx.execute(refactorQuery, Map.of("conf", Map.of())).stream())
+                    .singleElement(InstanceOfAssertFactories.map(String.class, Object.class))
+                    .hasEntrySatisfying("error", e -> assertThat(e).asString().contains("EntityNotFoundException"));
+            tx.rollback();
+        }
     }
 
     private void issue2797Common(String extractQuery) {

--- a/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
+++ b/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
@@ -1962,6 +1962,42 @@ public class GraphRefactoringTest {
         }
     }
 
+    @Test
+    public void mergeNodesWithConstraints() {
+        db.executeTransactionally("CREATE CONSTRAINT foo_uniq FOR ()-[r:MY_REL]-() REQUIRE r.foo IS UNIQUE");
+        db.executeTransactionally(
+                """
+                   CREATE
+                       (n1 {name: "n1"})-[r1:MY_REL {foo: "a"}]->(n2 {name: "n2"}),
+                       (n3 {name: "n3"})-[r2:MY_REL {foo: "b"}]->(n4 {name: "n4"})""");
+        db.executeTransactionally("CALL db.awaitIndexes()");
+        db.executeTransactionally(
+                """
+                   MATCH (n1 {name: "n1"}), (n3 {name: "n3"})
+                   CALL apoc.refactor.mergeNodes([n1, n3], {properties: 'discard'}) YIELD node
+                   FINISH""");
+
+        try (final var tx = db.beginTx()) {
+            final var query =
+                    """
+                    MATCH (a)
+                    OPTIONAL MATCH (a)-[r]->(b)
+                    RETURN
+                      a.name,
+                      CASE WHEN r.foo IS NULL THEN 'null' ELSE r.foo END AS `r.foo`,
+                      CASE WHEN b.name IS NULL THEN 'null' ELSE b.name END AS `b.name`
+                    ORDER BY `a.name`, `r.foo`, `b.name`
+                    """;
+            assertThat(tx.execute(query).stream())
+                    .containsExactly(
+                            Map.of("a.name", "n1", "r.foo", "a", "b.name", "n2"),
+                            Map.of("a.name", "n1", "r.foo", "b", "b.name", "n4"),
+                            Map.of("a.name", "n2", "r.foo", "null", "b.name", "null"),
+                            Map.of("a.name", "n4", "r.foo", "null", "b.name", "null"));
+            tx.commit();
+        }
+    }
+
     private void issue2797Common(String extractQuery) {
         db.executeTransactionally("CREATE CONSTRAINT unique_book FOR (book:MyBook) REQUIRE book.name IS UNIQUE");
         db.executeTransactionally("CREATE (n:MyBook {name: 1})");

--- a/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
+++ b/core/src/test/java/apoc/refactor/GraphRefactoringTest.java
@@ -1807,6 +1807,79 @@ public class GraphRefactoringTest {
         }
     }
 
+    @Test
+    public void invertWithConstraints() {
+        db.executeTransactionally("CREATE CONSTRAINT id_unique FOR ()-[r:R]-() REQUIRE r.id IS UNIQUE");
+        db.executeTransactionally("CREATE (a:A {id:'A'})-[r:R {id:'R'}]->(b:B {id:'B'})");
+        db.executeTransactionally("CALL db.awaitIndexes()");
+
+        try (final var tx = db.beginTx()) {
+            assertThat(tx.execute("MATCH ()-[r:R]->() CALL apoc.refactor.invert(r) YIELD error RETURN error").stream())
+                    .singleElement(InstanceOfAssertFactories.map(String.class, Object.class))
+                    .containsEntry("error", null);
+            tx.commit();
+        }
+        try (final var tx = db.beginTx()) {
+            final var query =
+                    """
+                    MATCH (a)
+                    OPTIONAL MATCH (a)-[r]->(b)
+                    RETURN
+                      a.id,
+                      CASE WHEN r.id IS NULL THEN 'null' ELSE r.id END AS `r.id`,
+                      CASE WHEN b.id IS NULL THEN 'null' ELSE b.id END AS `b.id`
+                    ORDER BY `a.id`, `r.id`, `b.id`
+                    """;
+            assertThat(tx.execute(query).stream())
+                    .containsExactly(
+                            Map.of("a.id", "A", "r.id", "null", "b.id", "null"),
+                            Map.of("a.id", "B", "r.id", "R", "b.id", "A"));
+            tx.commit();
+        }
+    }
+
+    @Test
+    public void invertErrorHandling() {
+        db.executeTransactionally("CREATE (a:A {id:'A'})-[r:R {id:'R'}]->(b:B {id:'B'})");
+
+        final var invertQuery =
+                "MATCH ()-[r:R]->() DELETE r WITH r CALL apoc.refactor.invert(r, $conf) YIELD error RETURN error";
+        try (final var tx = db.beginTx()) {
+            final var params = Map.<String, Object>of("conf", Map.of("failOnErrors", true));
+            assertThatThrownBy(() -> tx.execute(invertQuery, params).resultAsString())
+                    .hasMessageContaining(
+                            "Failed to invoke procedure `apoc.refactor.invert`: Caused by: org.neo4j.internal.kernel.api.exceptions.EntityNotFoundException")
+                    .hasRootCauseExactlyInstanceOf(
+                            org.neo4j.internal.kernel.api.exceptions.EntityNotFoundException.class);
+            tx.rollback();
+        }
+
+        try (final var tx = db.beginTx()) {
+            final var query =
+                    """
+                    MATCH (a)
+                    OPTIONAL MATCH (a)-[r]->(b)
+                    RETURN
+                      a.id,
+                      CASE WHEN r.id IS NULL THEN 'null' ELSE r.id END AS `r.id`,
+                      CASE WHEN b.id IS NULL THEN 'null' ELSE b.id END AS `b.id`
+                    ORDER BY `a.id`, `r.id`, `b.id`
+                    """;
+            assertThat(tx.execute(query).stream())
+                    .containsExactly(
+                            Map.of("a.id", "A", "r.id", "R", "b.id", "B"),
+                            Map.of("a.id", "B", "r.id", "null", "b.id", "null"));
+            tx.commit();
+        }
+
+        try (final var tx = db.beginTx()) {
+            assertThat(tx.execute(invertQuery, Map.of("conf", Map.of())).stream())
+                    .singleElement(InstanceOfAssertFactories.map(String.class, Object.class))
+                    .hasEntrySatisfying("error", e -> assertThat(e).asString().contains("EntityNotFoundException"));
+            tx.rollback();
+        }
+    }
+
     private void issue2797Common(String extractQuery) {
         db.executeTransactionally("CREATE CONSTRAINT unique_book FOR (book:MyBook) REQUIRE book.name IS UNIQUE");
         db.executeTransactionally("CREATE (n:MyBook {name: 1})");

--- a/core/src/test/resources/procedures/common/procedures.json
+++ b/core/src/test/resources/procedures/common/procedures.json
@@ -6138,7 +6138,7 @@
   } ]
 }, {
   "isDeprecated" : false,
-  "signature" : "apoc.refactor.from(rel :: RELATIONSHIP, newNode :: NODE) :: (input :: INTEGER, output :: RELATIONSHIP, error :: STRING)",
+  "signature" : "apoc.refactor.from(rel :: RELATIONSHIP, newNode :: NODE, config = {} :: MAP) :: (input :: INTEGER, output :: RELATIONSHIP, error :: STRING)",
   "name" : "apoc.refactor.from",
   "description" : "Redirects the given `RELATIONSHIP` to the given start `NODE`.",
   "returnDescription" : [ {
@@ -6168,10 +6168,16 @@
     "description" : "The node to redirect the given relationship to.",
     "isDeprecated" : false,
     "type" : "NODE"
+  }, {
+    "name" : "config",
+    "description" : "{\n    failOnErrors = false :: BOOLEAN\n}\n",
+    "isDeprecated" : false,
+    "default" : "DefaultParameterValue{value={}, type=MAP}",
+    "type" : "MAP"
   } ]
 }, {
   "isDeprecated" : false,
-  "signature" : "apoc.refactor.invert(rel :: RELATIONSHIP) :: (input :: INTEGER, output :: RELATIONSHIP, error :: STRING)",
+  "signature" : "apoc.refactor.invert(rel :: RELATIONSHIP, config = {} :: MAP) :: (input :: INTEGER, output :: RELATIONSHIP, error :: STRING)",
   "name" : "apoc.refactor.invert",
   "description" : "Inverts the direction of the given `RELATIONSHIP`.",
   "returnDescription" : [ {
@@ -6196,6 +6202,12 @@
     "description" : "The relationship to reverse.",
     "isDeprecated" : false,
     "type" : "RELATIONSHIP"
+  }, {
+    "name" : "config",
+    "description" : "{\n    failOnErrors = false :: BOOLEAN\n}\n",
+    "isDeprecated" : false,
+    "default" : "DefaultParameterValue{value={}, type=MAP}",
+    "type" : "MAP"
   } ]
 }, {
   "isDeprecated" : false,
@@ -6662,7 +6674,7 @@
   } ]
 }, {
   "isDeprecated" : false,
-  "signature" : "apoc.refactor.to(rel :: RELATIONSHIP, endNode :: NODE) :: (input :: INTEGER, output :: RELATIONSHIP, error :: STRING)",
+  "signature" : "apoc.refactor.to(rel :: RELATIONSHIP, endNode :: NODE, config = {} :: MAP) :: (input :: INTEGER, output :: RELATIONSHIP, error :: STRING)",
   "name" : "apoc.refactor.to",
   "description" : "Redirects the given `RELATIONSHIP` to the given end `NODE`.",
   "returnDescription" : [ {
@@ -6692,6 +6704,12 @@
     "description" : "The new end node the relationship should point to.",
     "isDeprecated" : false,
     "type" : "NODE"
+  }, {
+    "name" : "config",
+    "description" : "{\n    failOnErrors = false :: BOOLEAN\n}\n",
+    "isDeprecated" : false,
+    "default" : "DefaultParameterValue{value={}, type=MAP}",
+    "type" : "MAP"
   } ]
 }, {
   "isDeprecated" : false,

